### ISSUE TITLE
Increased mouse input lag with Moonlight 6.0.0

### ISSUE THREAD

**Issue by harm-less** [moonlight-stream/moonlight-qt#1321]
Yesterday I'd upgraded to the latest Moonlight and Sunshine, yet when I started streaming my desktop like I usually do the mouse input lag was noticeably worse. I'd say around 3 - 4 times as bad (at around 50 - 100 ms)

The "Optimize mouse for remote desktop" was turned on before and after the upgrade.

When I reverted back to 5.0.1 of Moonlight (my previous version) everything started working as I expected as there was only a slight and usual delay I was used to from before the upgrade.

I don't really have any logs or exact numbers atm, but I wanted to share to see if other people have the issue. If so I can do some follow up tests and document it better as it's currently working for me.

**Client PC details (please complete the following information)**
- OS: Windows 10 Pro (10.0.19043)
- Moonlight: 6.0.0
- GPU: Intel UHD Graphics 605
- Driver version: 26.20.100.7463

**Server PC details**
 - OS: Windows 11 Pro (10.0.19043)
 - Sunshine or GeForce Experience version: Sunshine v0.23.1
- GPU: RTX 4070 Ti
- Driver: 31.0.15.4665

**Comment by kostrykin** [comment]
Not 100% sure, whether this is related, but: I upgraded to Moonlight 6 yesterday too, and my "average rendering time" has increased from 0.1-0.2 ms to 2-3 ms. It indeed feels like increased input lag, but judging by the stats, it's the rendering time.

My client OS is macOS.

**Comment by makedir** [comment]
It is a bit annoying that the developer of this app randomly switches the way the video is decoded all the time instead of giving the user the option to decide which way it is decoded (dx9, dx11, dvxa and so on), at least under Windows. I also notice there is some weird bug randomly going on with input lag when you chose fullscreen and also use VSYNC, the render latency sometiems is great around 0.8ms but then randomly is bad around 2-3ms and switching out and back into the decode fullscreen video sometimes fixes it, but not always.

**Comment by Pointyguide** [comment]
Yes average rendering time increased with v6

**Comment by kostrykin** [comment]
> Yes average rendering time increased with v6

What is your client OS? @Pointyguide

**Comment by Pointyguide** [comment]
> > Yes average rendering time increased with v6
> 
> What is your client OS? @Pointyguide

Macos 14.5 M2 mac mini.

macOS has never had as good of latency feeling as a windows client.

**Comment by cgutman** [comment]
> I don't really have any logs or exact numbers atm, but I wanted to share to see if other people have the issue. If so I can do some follow up tests and document it better as it's currently working for me.

I suspect this is #1304 (which is also affecting an Intel GPU of a similar generation)

Answering a few questions about your Moonlight config would help:
- Client CPU model
- Resolution, frame rate, and bitrate set in Moonlight
- Full-screen, borderless windowed, or regular windowed mode?
- V-sync on or off?
- HDR on or off?

Please do follow up with logs too (from both 5.0.1 and 6.0.0 would be ideal).

> It is a bit annoying that the developer of this app randomly switches the way the video is decoded all the time instead of giving the user the option to decide which way it is decoded (dx9, dx11, dvxa and so on), at least under Windows.

You have always been able to choose via environment variables using`DXVA2_ENABLED=0` or `D3D11VA_ENABLED=0` to disable DXVA2 and D3D11VA respectively.

I'm not just switching things for no reason, nor is writing a completely new renderer from scratch easy. D3D11VA is there (and the new default) because it performs significantly better in many situations, and the old DXVA2 renderer has poor support and encounters many driver bugs on newer hardware. I've been easing in the D3D11VA renderer over many releases precisely to avoid widespread breakage.

> I also notice there is some weird bug randomly going on with input lag when you chose fullscreen and also use VSYNC, the render latency sometiems is great around 0.8ms but then randomly is bad around 2-3ms and switching out and back into the decode fullscreen video sometimes fixes it, but not always.

Depending on your settings that is probably normal (or at least the numbers aren't the problem). In general just staring at these statistics changing a couple milliseconds one way or another isn't really productive or indicative of a good or bad streaming experience. If you can perceive it strongly enough to file a bug for it, it's _not_ a 1-2 ms change you're feeling - it's something else much larger.

For detailed analysis, you may find it useful to use a tool like [PresentMon](https://github.com/GameTechDev/PresentMon) or [System Informer](https://systeminformer.sourceforge.io/downloads.php) to watch the state of the graphics subsystem in real-time. That will show you the actual display latency of each frame, which Moonlight can't measure itself. It's possible that the display latency is the same (or even that it goes _down_) when you see slightly higher rendering times.

> Not 100% sure, whether this is related, but: I upgraded to Moonlight 6 yesterday too, and my "average rendering time" has increased from 0.1-0.2 ms to 2-3 ms. It indeed feels like increased input lag, but judging by the stats, it's the rendering time.

This is unexpected as most macOS users have reported _better_ latency with the new Metal renderer than the old one (#1087 and various reports on Discord). Please split this into a separate ticket and include details about your macOS version, the model of your Mac, and what settings (resolution, frame rate, bitrate, and v-sync state) you're using to stream. The Mac and Windows renderers are completely different code, so any bugs between them are not related to each other.

Also understand that the new renderer is implemented using a lower-level framework (Metal), so it will observe (and therefore count) more of the display pipeline latency than the old renderer did. That means the rendering latency numbers are not directly comparable between the two.

**Comment by serdargitthub** [comment]
@cgutman you mentioned that we are able to choose via environment variables usingDXVA2_ENABLED=0 or D3D11VA_ENABLED=0 to disable DXVA2 and D3D11VA respectively.

can you please describe where to put these vars ? thank you

**Comment by cgutman** [comment]
@serdargitthub they are environment variables, so you can set them in a Command Prompt window. For example:
```
set D3D11VA_ENABLED=0
"C:\Program Files\Moonlight Game Streaming\Moonlight.exe"
```

**Comment by makedir** [comment]
@cgutman please give drop down menu inside moonlight for windows so you can chose dx9, dx11, also more options for decoder like dxva or other. also put these information being used in the overlay.

**Comment by cgutman** [comment]
> please give drop down menu inside moonlight for windows so you can chose dx9, dx11, also more options for decoder like dxva or other. also put these information being used in the overlay.

The DX9 renderer will probably be removed in Moonlight v6.1 when I remove support for Windows 7/8.1 since all supported systems will support DX11.

Putting the renderer name in the overlay is a good idea though.

As for the original report in this bug, I have a root cause and fix available for testing in #1304. I'm going to close this as a duplicate of #1304.
